KEP-4578: Server Feature Gate in etcd (#4610)
kubernetes/enhancements: keps/sig-etcd/4578-feature-gate/README.md

When it is time to [graduate an experimental feature to stable](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/features.md#graduating-an-experimental-feature-to-stable), a new stable feature flag identical to the experimental feature flag but without the --experimental prefix is added to replace the old feature flag, which is a breaking change.

We are proposing to add [feature gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) in the similar way as Kubernetes does, so that can turn on or off any feature using the `--feature-gates` flag.

Benefits from this enhancement includes:

* graduating a feature is no longer a breaking change with new flags.

* well-established feature lifecycle convention from Kubernetes.

* a unified place to store all the feature enablement state in the code base for simplicity, and easy to refer to in guarding the implementation code.

### Goals

* Introduce the `--feature-gates` flag in etcd code base.

> * Introduce apis/methods to query the enablement state of a feature.

Reviewer: Should we also be including a reference implementation for querying these new APIs via `etcdctl`?  If so let's be clear this is in scope.

Author: yes, we do plan to add `etcdctl` commands on top of the grpc support. Made it explicit in the KEP.

Reviewer 2: Should we consider exporting it as a metric? Kubernetes has already done this. Metrics can be used to easily query which etcd has enabled a certain feature.

```shell
$ curl -X GET localhost:8001/metrics|grep feature
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0# HELP kubernetes_feature_enabled [BETA] This metric records the data about the stage and enablement of a k8s feature.
# TYPE kubernetes_feature_enabled gauge
kubernetes_feature_enabled{name="APIListChunking",stage=""} 1
kubernetes_feature_enabled{name="APIPriorityAndFairness",stage=""} 1
kubernetes_feature_enabled{name="APIResponseCompression",stage="BETA"} 1
kubernetes_feature_enabled{name="APISelfSubjectReview",stage=""} 1
kubernetes_feature_enabled{name="APIServerIdentity",stage="BETA"} 1
kubernetes_feature_enabled{name="APIServerTracing",stage="BETA"} 1
kubernetes_feature_enabled{name="AdmissionWebhookMatchConditions",stage="BETA"} 1
kubernetes_feature_enabled{name="AggregatedDiscoveryEndpoint",stage="BETA"} 1
...
```

Author: That's a good idea! Thanks for bringing it up!

* `{server_endpoint}/featuregate?feature={featureName}` returns true if the feature is enabled for the cluster
* `{server_endpoint}/featuregate?feature={featureName}&isServerFeature=true` returns true if the feature is enabled for the server

### Feature Gate

We will use the new `k8s.io/component-base/featuregate.VersionedSpecs`(introduced in [kep-4330](https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/4330-compatibility-versions)) to register and track the features along with their lifecycle at different release versions. 

```go
map[Feature]VersionedSpecs {
		featureA: VersionedSpecs{
			{Version: mustParseVersion("1.27"), Default: false, PreRelease: Beta},
			{Version: mustParseVersion("1.28"), Default: true, PreRelease: GA},
		},
		featureB: VersionedSpecs{
			{Version: mustParseVersion("1.28"), Default: false, PreRelease: Alpha},
		},
		featureC: VersionedSpecs{
			{Version: mustParseVersion("1.28"), Default: false, PreRelease: Beta},
		},
		featureD: VersionedSpecs{
			{Version: mustParseVersion("1.26"), Default: false, PreRelease: Alpha},
			{Version: mustParseVersion("1.28"), Default: true, PreRelease: Deprecated},
		}
}
```

> The feature gates for a server can only be set with the `--feature-gates` flag during startup. We do not support dynamically changing the feature gates when the server is running.

Reviewer: Should we support setting feature flags via configuration file as well?  Imo yes as many of our users prefer to configure etcd via configuration file.

Benefits from this enhancement includes:

* graduating a feature is no longer a breaking change with new flags.

* well-established feature lifecycle convention from Kubernetes.

* a unified place to store all the feature enablement state in the code base for simplicity, and easy to refer to in guarding the implementation code.

### Goals

* Introduce the `--feature-gates` flag in etcd code base.

* Introduce apis/methods to query the enablement state of a feature.

* Establish lifecycle stages of a feature, and clear criteria of lifecycle progression.

### Non-Goals

* Migrate all existing `--experimental` features to feature gate.

* Use feature gate as a mechanism to fix bugs.

## Proposal

### User Stories (Optional)

#### Story 1

It would be easier for developers to add new features without worrying too much about breaking etcd, because the feature would be Alpha first and disabled by default, and new changes would be mostly gated behind the feature gate.

> It would be a smoother process to move the feature to the next stage without the need to introduce new flags.

Reviewer 3: Adding feature gate and introducing new flags should be orthogonal? Enabling a feature gate may enable users to configure certain flags for the feature.

Author: Rephrase it a little bit. Can you check if it is more clear?

There are also features like `ExperimentalEnableDistributedTracing` (enables distributed tracing using OpenTelemetry protocol), which could work fine at the local server level.

We are proposing to have separate APIs to query if a feature is enabled for the local server, and for the whole cluster.

#### Should we use feature gate for bug fixes?

There are some use cases like `ProgressNotify`, for which some bugs are found and later fixed in a patch version. The client would need to know if the etcd version in use contains that fix to decide whether or not to use that feature. 

The question is: should a new feature gate be added to signal the bug fix? 

We think the answer is NO:
* the new feature would need to be enabled by default to always apply the bug fix for new releases.
* it changes the API which is not desirable in patch version releases.

The proper way of handling cases like `ProgressNotify` is:
1. the feature should be gated by the feature gate from the beginning.
1. the feature should be disabled by default until it is widely tested in practice. 

> 1. when the bus is found, the feature should ideally be at a lifecycle in which it is disabled by default. If not, the admin should disable it by the `--feature-gates` flag.

Reviewer 3: ```suggestion
1. when the bug is found, the feature should ideally be at a lifecycle in which it is disabled by default. If not, the admin should disable it by the `--feature-gates` flag.
```

We think we should still stick to the common practice of not changing the lifecycle of a feature in patch versions. Because:
* changing the lifecycle of a feature is an API change. According to the [etcd Operations Guide](https://etcd.io/docs/v3.5/op-guide/versioning/), only new minor versions may add additional features to the API.
* bugs in etcd could be hard to detect, and the reliability and robustness of etcd is more important than speed. A long history of testing through practical adoption a new feature is beneficial.

With the feature gate in place, we could consider increasing the etcd release cadence because it would be easier to add new features and less risky to release new features.

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### New APIs

> New `--feature-gates` flag would be added to start the etcd server, with format like `--feature-gates=featureA=true,featureB=false`. The flag can also be set in the `config-file`.

Reviewer 4: (nit)
```suggestion
A new `--feature-gates` command line argument would be added to start the etcd server, with format like `--feature-gates=featureA=true,featureB=false`. The flag can also be set in the `config-file`.
```

> New grpc and http endpoints would be added to query if a feature is enabled for the server or cluster. New `etcdctl` will also be added.

Reviewer 3: ```suggestion
New grpc and http APIs would be added to query if a feature is enabled for the server or cluster. New `etcdctl` command will also be added.
```

For grpc, new rpc type would be added to the `Maintenance` service.

```proto
service Maintenance {
  ...
  rpc FeatureGateStatus(FeatureGateStatusRequest) returns (FeatureGateStatusResponse) {
    option (google.api.http) = {
      post: "/v3/maintenance/featuregate"
      body: "*"
    };
  }
  ...
}
message Feature {
  string name = 1;
  bool enabled = 2;
}

message FeatureGateStatusRequest {

>   // if true, query if the features are enabled for the single server.
>   // otherwise, query if the features are enabled for the cluster.
>   bool isServerFeature = 1;

Reviewer 3: For simplicity, suggest not to add the concept of "cluster level feature gate". If inconsistency across all nodes in the cluster may cause problem, then let's clearly clarify it in the document, and print warning/error in log if possible. If the inconsistency may cause critical issue (i.e. data inconsistency), then we can even fail the etcd's startup (but I do not see such case so far).

Also this is a generic issue, not limited to only feature gate. etcd has lots of flags, the inconsistencies across all nodes for each flag may cause issue or confusion.
```suggestion
```

Author: inconsistencies is inevitable in upgrade/downgrade scenarios. For example a feature can be alpha (disabled by default) in one version, and beta (enabled by default) in the next. I think we have to have a clear way of handling the inconsistencies. Just warning/error messages is not enough.

Reviewer 3: - Name `clusterLevelFeature` makes more sense then `isServerFeature`. But see bigger concern below
- With respect to the inconsistence, 
  - As mentioned above, inconsistence is a generic issue when upgrading/updating a cluster, not only limited to the feature gate. For example, when we change the value of a flag (it can be any flag, including the feature gate) during upgrading or updating a cluster, there is a small window of inconsistence. So if we really want to resolve such issue, it's out of the scope of feature gate, we should have a more generic solution.
  - But I don't see any major or critical issue for such inconsistence. So we need to have a list of such issues firstly, and clearly clarify each issue, then discuss the solution. It can be discussed separately.

Author: feature gate is a potential solution to inconsistencies. 
* if a flag is available in both versions, it is probably ok to assume they are consistently set during upgrade/downgrade.
* if a flag is available only in one version, ideally it should be gated behind a feature gate, and feature gate would gracefully handle the case when it is set in one version but not in the other. 

Let's imagine a hypothetical case:
1. a new write type is introduced in V_n+1, and some K8s function depends on this new feature.
2. K8s asks an etcd server (which happens to be V_n+1) if the feature is enabled, and gets confirmation, and starts sending the new write requests.
3. the new write requests would be sent to etcd leader (which happens to be V_n), and it does not recognize this new write type. 

However short this intermediate mixed version state is, the workload could potentially break the whole cluster. Right now, we have to introduce a new raft type in V_n before using it in V_n+1 to prevent such failures. But if we can handle cluster feature enablement, we could remove the step of introducing a new raft type in V_n. 
I think there is great benefit in speeding up the dev process with this cluster level feature gate.

Author: Another example is the ProgressNotify, which [KEP-2340](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md) depends on. So K8s performance can be greatly impacted by whether this feature is available or not. If there is no cluster level feature information, depending on which api server is the master, or which etcd server it talks to, ProgressNotify could flip flop, and making the performance inconsistent. There is no guarantee etcd upgrade/downgrade could be done in x seconds, I think we should make our best to make the intermediate state predictable. 

cc @serathius

Reviewer 5: Cluster level flags are not outside the scope of the KEP. Let's get back to the discussion when the new proposal will be ready for review.

We choose to store the `enabled_features` through raft instead of dynamically querying the `\featuregate` endpoint because
1. it saves network bandwidth.
1. we want the cluster feature enablement status to be as stable as possible, only subject to change when the admin intends it as in the case of adding/removing/upgrading/downgrading/restarting servers. In the case of unpredictable loss of a member server, the feature gate of the cluster should not change.

In `membership.RaftCluster`, we will add a new an aggregate method `FeatureEnabled(key Feature) bool` to query cluster level enablement of a feature. 

```go
// A feature is enabled for the whole cluster only if it is enabled for all the members. 
// In case of a conflict, the feature would be disabled, and would not bring down the whole cluster because conflict is expected during upgrade/downgrade.
func (*RaftCluster) FeatureEnabled(key Feature) bool {}
```

### Feature Stages

> Following the convention of Kubernetes, a feature can go through a lifecycle of Alpha -> Beta -> GA -> Deprecated. 

Reviewer 4: (nit)
```suggestion
Following the convention of Kubernetes, a feature can go through a lifecycle of Alpha → Beta → GA → Deprecated.
```

##### Prerequisite testing updates

##### Unit tests

New feature gate unit tests will be added.

##### Integration tests

New feature gate integration tests will be added.

##### e2e tests

We will add one cluster level and one server level experimental features to the feature gate (without removing the original experimental flag), and add e2e tests to make sure the feature gate functions equivalently to their `--experimental-xxx` flags.

We will also add downgrade/upgrade e2e tests to make sure the feature gate does not break the cluster in mixed version cluster.

### Graduation Criteria

#### Milestone 1

* server level feature gate implemented.
* cluster level feature gate implemented.

#### Milestone 2

* backport necessary proto changes to 3.5.
* http endpoints added.
* grpc endpoints added.
* `etcdctl` command added.
* equivalent of one cluster level and one server level experimental features added to the feature gate.
* initial e2e tests completed.

#### Milestone 3

* e2e tests for upgrade & downgrade & mixed version scenarios completed.
* robustness tests completed. 
* documentation.

### Upgrade / Downgrade Strategy

The feature gate feature would available in 3.6+. 

> In 3.5, the necessary proto changes would be backported, but feature gate would not be available. 

Reviewer 3: Why do we need to backport proto change to 3.5? We are just adding the feature of feature gate. We do not add any new feature which may change or break the proto.

For the impact on the upgrade/downgrade, we need to analyze each new feature case by case. But for this feature (feature-gate), it seems safe for now.

If we are planning to migrate some existing experimental feature, we can do it in 3.7.

Author: The plan is for the cluster level features, each member needs to know the enabled features in other members, and save that in a new member Attribute filed.

Reviewer 3: Let's try not impact 3.5 in any way. We should keep the downgrade as simple & easy as possible.

FYI. I am thinking to simplify the existing downgrade workflow. But it can be discussed separately.

Author: Added the `Risks and Mitigations` section to explain this change should not really impact 3.5. Does it make sense @ahrtr ?

Reviewer 5: No longer in scope.

### Version Skew Strategy

Since we are introducing new fields in `member.Attributes` which is sent through raft, we will backport the new fields into 3.5, but would not use them in any way affecting 3.5 etcd server.

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

<!--
What other approaches did you consider, and why did you rule them out? These do
not need to be as detailed as the proposal, but should include enough
information to express the idea and why it was not acceptable.

> -->

Reviewer 4: We could introduce the internal feature gate support in a minor version of etcd (eg 3.6), so that an 3.6 etcd can see what feature gates are active. And then release etcd 3.7 that actually lets you turn on feature gates.

I think that would be more effort but a better downgrade story.

Author: etcd has very long release cycles. If possible, it would be good to have the feature gate available in 3.6. That's why we are proposing to have minimal backport in 3.5, so that it is ok to start using them in 3.6.

checklist items _must_ be updated for the enhancement to be released.
-->

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
- [ ] (R) KEP approvers have approved the KEP status as `implementable`
- [ ] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place
- [ ] (R) Graduation criteria is in place
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [etcd-io/website], for publication to [etcd.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[etcd.io]: https://etcd.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[etcd-io/website]: https://github.com/etcd-io/website

## Summary

> We are introducing a new `--feature-gates` flag in etcd, and the underlying framework to gate future feature enhancement to etcd. 

Reviewer 5: Adding a flag should not be a goal of it's self, but a means. Please describe what you are proposing in one paragraph.

Author: Thanks for the suggestion. Updated the summary, can you take a look again?

Reviewer 5: Looks great, thank you!

We are proposing to add [feature gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) in the similar way as Kubernetes does, so that can turn on or off any feature using the `--feature-gates` flag.

Benefits from this enhancement includes:

* graduating a feature is no longer a breaking change with new flags.

* well-established feature lifecycle convention from Kubernetes.

* a unified place to store all the feature enablement state in the code base for simplicity, and easy to refer to in guarding the implementation code.

* easy to query if a feature is available for an ectd cluster.

### Goals

* Introduce the `--feature-gates` flag in etcd code base.

* Introduce apis/methods to query the enablement state of a feature.

* Establish lifecycle stages of a feature, and clear criteria of lifecycle progression.

### Non-Goals

> * Migrate all existing `--experimental` features to feature gate.

Reviewer 3: We should clearly clarify how we will handle the existing `--experimental` features.

Reviewer 6: My understanding is that the existing experimental feature will not be different unless they are migrated feature gate. @siyuanfoundation is my understanding correct? Maybe the action item here is to clarify that if that's the case to avoid confusion like this.

Author: added more explanation. @ahrtr @wenjiaswe is it more clear?

Reviewer 6: The updated non-goals look pretty clear to me. Thanks!

It might also worth mentioning somewhere regarding what happens to the old "--experimental" flag after the feature gate implementation, after the migration, if not stated in the doc yet. I assume it will continue working as before, but might worth specifying that in case people has concern

Author: Thanks @wenjiaswe for the suggestion. I think it is covered by the `Path to Migrate Existing Experimental Features` section.

Reviewer 6: Ah, sorry I missed that, thanks for pointing it out. 

I don't have any question now:) Thanks!

When it is time to [graduate an experimental feature to stable](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/features.md#graduating-an-experimental-feature-to-stable), a new stable feature flag identical to the experimental feature flag but without the --experimental prefix is added to replace the old feature flag, which is a breaking change.

We are proposing to add [feature gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) in the similar way as Kubernetes does, so that can turn on or off any feature using the `--feature-gates` flag.

Benefits from this enhancement includes:

* graduating a feature is no longer a breaking change with new flags.

* well-established feature lifecycle convention from Kubernetes.

* a unified place to store all the feature enablement state in the code base for simplicity, and easy to refer to in guarding the implementation code.

* easy to query if a feature is available for an ectd cluster.

### Goals

* Introduce the `--feature-gates` flag in etcd code base.

* Introduce apis/methods to query the enablement state of a feature.

> * Establish lifecycle stages of a feature, and clear criteria of lifecycle progression.

Reviewer 7: Since we also cover consensus-based feature enablement in this KEP, can we add that as an explicit goal here? Say `Establish a mechanism to safely and consistently enable/disable cluster-level features`

Author: Thanks @shyamjvs. These are all very good points. After some thoughts, we decided to move cluster level feature gate into another KEP, we can continue the discussions there https://github.com/kubernetes/enhancements/issues/4647

Reviewer 7: +1 to that - will surely make this KEP more tractable.

One question - is there any coupling we expect between member-level and cluster-level features (like the APIs/mechanisms to enable and expose them)? If so, it'll help to roughly address in this KEP how we're leaving space to address the latter usecase in future.

Author: In this KEP, there is a line of `we will not expose querying feature enablement of any cluster level features, because we cannot risk possible conflicting settings that might break cluster correctness.`. 
The alternative is to expose the cluster level features as per server feature, and users can aggregate them across all members, but that could be prone to misusage like when to use aggregation, and not really based on a raft consensus.

Reviewer 7: yeah verifiability and correctness will be hard to prove from client-side imho (similar to how `/members` API gives consistent list of members rather than clients assuming it to be "running instances")

Reviewer 5: > The alternative is to expose the cluster level features as per server feature, and users can aggregate them across all members, but that could be prone to misusage like when to use aggregation, and not really based on a raft consensus.

Don't. Let's follow up the discussion when we have a KEP to propose a cluster level feature gates.

### Non-Goals

* Migrate all existing `--experimental` features to feature gate. 
  * While feature gate provides a [path to deprecate existing `--experimental` features]((#path-to-migrate-existing-experimental-features)), and we will migrate several `--experimental` during the feature gate implementation as examples, the completion of the KEP does not require migrating all existing `--experimental` features. It can be done as followup tasks.

* Use feature gate as a mechanism to fix bugs.

## Proposal

> ### User Stories (Optional)

Reviewer 7: One story that's unclear for me (not covered in any of the below so asking) - do we foresee feature gates that once enabled by the cluster, cannot (or should not) be disabled? For e.g because it would lead to breaking changes in the apply/storage layer. Depending on the feature, we may need an offline migration step or enforce the feature to be uni-directional or something else. But this is something we should deliberate on here imho.

FWIW this problem seems equivalent to version downgrade support - so cc @serathius for ideas/opinions.

Reviewer 3: >  For e.g because it would lead to breaking changes in the apply/storage layer

We should never allow etcd running into such situation, otherwise it may lead to data inconsistencies. 

> One story that's unclear for me (not covered in any of the below so asking) - do we foresee feature gates that once enabled by the cluster, cannot (or should not) be disabled? 

cluster-level feature isn't in the scope of this KEP, it's a complicated topic. But personally I think we should only allow users to disable a cluster-level feature **offline**.

Reviewer 5: Offline enablement doesn't guarantee consistency, as different members may be at various stages of applying the WAL. Therefore, changing cluster behavior necessitates a Raft entry. Let's close this discussion, as it's outside the scope of this KEP. For cluster-level flags, I'd like to be involved from the early stages of the KEP and be extremely meticulous about potential correctness issues.

#### Story 1

It would be easier for developers to add new features without worrying too much about breaking etcd, because the feature would be Alpha first and disabled by default, and new changes would be mostly gated behind the feature gate.

It would be a smoother process to move the feature to the next stage without the need to introduce a potentially breaking change of an `--experimental` flag.

The developer also would not need to worry about feature compatibility when there is a mixed version cluster.

#### Story 2

Users of etcd cluster would be able to find out which features are enabled in the etcd cluster and decide on how to use them downstream. 

Currently Kubernetes uses [`FeatureSupportChecker`](https://github.com/kubernetes/kubernetes/blob/db82fd1604ebf327ab74cde0a7158a8d95d46202/staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature/feature_support_checker.go#L42) to check if a feature is supported by etcd, which basically checks the etcd version and compares the version with a hard-coded map of feature availability for different etcd versions.

#### Story 3

> In a HA cluster, users should be able to enable/disable a feature without bring down the whole cluster, by restarting the servers with the new feature gate flag one by one. 

Reviewer 7: Given cluster-level etcd features decouple "when a feature gate gets turned on for member(s)" and "when the feature actually gets started to be used by the cluster" - I'd recommend completing this story by adding `.. And for cluster-level features, be able to query whether the feature is currently in use by the cluster`.

Reviewer 7: And for debuggability, we might need to go a step further and also return `when` the feature was enabled/disabled (say the `consistent_index` where cluster agreed to do that).

Reviewer 5: Cluster level flags are not outside the scope of the KEP.

#### Story 4

> During cluster upgrade/downgrade, feature changes across versions should have predictable behavior in this mix version scenario. 

Reviewer 7: Can we assert a stronger claim/invariant here, say - `predictable behavior governed by the lowest member version`? Using concrete stories early on in the KEP can improve the mental model for what's proposed later on.

Reviewer 7: Also if there's a good reason why we don't want to be that specific here, open to hearing it :)

Reviewer 5: Cluster level features were dropped from the KEP scope, no longer need to consider upgrade/downgrade.

There are features like `ExperimentalEnableLeaseCheckpoint` (enables leader to send regular checkpoints to other members to prevent reset of remaining TTL on leader change), if different server nodes have different enablement values for `ExperimentalEnableLeaseCheckpoint`, the results would be inconsistent depending on which one is the leader and confusing.

There are also features like `ExperimentalEnableDistributedTracing` (enables distributed tracing using OpenTelemetry protocol), which could work fine at the local server level.

We are proposing to have separate APIs to query if a feature is enabled for the local server, and for the whole cluster.

#### Should we use feature gate for bug fixes?

There are some use cases like `ProgressNotify`, for which some bugs are found and later fixed in a patch version. The client would need to know if the etcd version in use contains that fix to decide whether or not to use that feature. 

The question is: should a new feature gate be added to signal the bug fix? 

We think the answer is NO:
* the new feature would need to be enabled by default to always apply the bug fix for new releases.

> * it changes the API which is not desirable in patch version releases.

Reviewer 7: One exception to consider here is - a long-standing bug that can only be fixed by making a breaking API/semantic change. In some cases, the bug fix may be important enough to release in a patch version. To do that safely though, we'd likely need a feature gate. So we might wanna keep this door open.

A concrete example coming to my mind for the above is https://github.com/kubernetes/kubernetes/pull/123935.

Reviewer 7: I do realize this is more of an exception than the norm. Nonetheless, it feels prudent for us to keep that door open.
